Remove erroneous optimization from eqtype

goldfirere opened the pull request:
From the top of `Ctype.eqtype`, I've removed a call to `eq_type` and added this comment:

```
  (* It's tempting to check for physical equality via [eq_type] here, but that
     would be incorrect: imagine comparing ['a * 'a] with ['b * 'a]. The
     first ['a] and ['b] would be identified in [eqtype_subst], and then
     the second ['a] and ['a] would be [eq_type]. So we do not call [eq_type]
     here. *)
```

See if you agree with the comment.

There is no known actual bug that this fixes, and thus no test case. But I'm doing some other work around here in the Jane Street branch, and the check is causing problems in that context.

goldfirere committed:
Remove erroneous optimization from eqtype

Octachron commented on typing/ctype.ml:
```diff
@@ -4131,8 +4131,11 @@ let eqtype_subst type_pairs subst t1 t2 =
+     would be incorrect: imagine comparing ['a * 'a] with ['b * 'a]. The
+     first ['a] and ['b] would be identified in [eqtype_subst], and then
+     the second ['a] and ['a] would be [eq_type]. So we do not call [eq_type]
+     here. *)
```
There is another instance of `eq_type` in the `_` case of the pattern matching.

Octachron commented:
I agree that mixing the  equalities of `eqtype_subst` with the physical `eq_type` is only valid if the type variable nodes of both types are separated. Now, I am wondering if those type variables nodes can be shared in the vanilla compiler.

goldfirere committed:
Remove eq_type from more and fewer places.

goldfirere commented:
It turns out that the physical equality check is necessary when `rename` is false, as showed up when I tried to remove the further calls to `eq_type` that you found @Octachron. Hopefully this version is better.

goldfirere committed:
Changelog

Octachron commented:
We could also handle the variable case directly:
```ocaml
    | (Tvar _, Tvar _)  ->
        if rename then eqtype_subst type_pairs subst t1 t2
        else if eq_type t1 t2 then ()
        else raise_unexplained_for Equality
```
But I agree that using the representative physical equality is correct without renaming and faster.

Octachron commented on typing/ctype.ml:
```diff
@@ -4131,8 +4131,19 @@ let eqtype_subst type_pairs subst t1 t2 =
+  let check_phys_eq t1 t2 =
+    not rename && eq_type t1 t2
+  in
+  (* Checking for physical equality when [rename] is true
```
Maybe `physical equality of type representatives` to avoid giving the impression that `eq_type` is `==`?

Octachron commented:
I am also wondering if we want to add a short-circuit test in `Ctype.equal`:
```ocaml
let equal env rename tyl1 tyl2 =
  if not rename && List.for_all2 eq_type tyl1 tyl2 then () else
  let subst = ref [] in
  try eqtype_list rename (TypePairs.create 11) subst env tyl1 tyl2
  with Equality_trace trace ->
    raise (Equality (expand_to_equality_error env trace !subst))
```
under the hypothesis that the optimisation is safe and could trigger not that infrequently in this location.

goldfirere committed:
Propagate lack of separation assumption

If [eqtype] no longer assumes separation, then [Printtyp.Names]
shouldn't either.

goldfirere committed:
Optimization as suggested by Florian

Octachron commented on typing/ctype.ml:
```diff
@@ -4318,20 +4333,23 @@ and eqtype_row rename type_pairs subst env row1 row2 =
 let equal env rename tyl1 tyl2 =
+  if List.length tyl1 <> List.length tyl2 then
+    raise_unexplained_for Equality;
+  if not rename && List.for_all2 eq_type tyl1 tyl2 then () else
```
Since there are no substitution in scope at this point, the `not rename` guard should not be needed, isn't it? If `tyl1` and `tyl2` are physically equal, they will be equal with the trivial substitution.

goldfirere committed:
Improvement suggested by Florian

gasche commented on typing/ctype.ml:
```diff
@@ -4131,8 +4131,18 @@ let eqtype_subst type_pairs subst t1 t2 =
+     first ['a] and ['b] would be identified in [eqtype_subst], and then the
+     second ['a] and ['a] would be [eq_type]. So we do not call [eq_type] here.
+
+     On the other hand, when [rename] is false we need to check for phyiscal
```
```suggestion
     On the other hand, when [rename] is false we need to check for physical
```

gasche commented on typing/ctype.ml:
```diff
@@ -4157,7 +4167,7 @@ let rec eqtype rename type_pairs subst env t1 t2 =
           | (Tconstr (p1, tl1, _), Tconstr (p2, tl2, _))
                 when Path.same p1 p2 ->
-              eqtype_list rename type_pairs subst env tl1 tl2
+              eqtype_list_same_length rename type_pairs subst env tl1 tl2
```
(I'm not sure why we feel so confident that the two argument lists have the same length here.)

goldfirere replied:
This is actually something I've wondered about when working on the compiler: what invariants exist for `type_expr`s? For example, I *think* we have the invariant that the level of a `type_expr` is always greater than or equal to the level of any type accessible from it. What other invariants do we have? It would be good to document them.

One invariant I know we do not have is that a `Tconstr` satisfies constraints. This is because we sometimes apply a path to fresh variables and then unify in order to satisfy constraints. So we mostly satisfy constraints, but not quite always.

Yet I think, as I've read the compiler, that we do always have it such that the arity of a type constructor is respected in a `Tconstr`. If I've gotten this wrong in this patch, the result is not catastrophic: it would lead to an exception being thrown in an erroneous program. That's a terrible user experience, but it is not unsound.

In any case, I'm content to remove this tiny optimization if others are made nervous by it.

Octachron replied:
I agree that the invariant `"Tconstr"s have consistent arity` is already expected in many parts of the compiler, it seems fine to me to use it there too.

garrigue commented:
After the explanation, I think I understand the problem, and why it doesn't happen in practice.
There are two ways to use `eqtype`: either to check equivalent types during module subtyping, allowing renaming, but using disjoint types, or to check type equality in a context where you don't want to modify types like unify would, and not allowing renaming. But you are right that you are not losing any useful optimization, and it is always better to have correct code without assuming implicit invariants.

goldfirere committed:
Typo (spotted by @gasche)

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

Octachron commented on typing/printtyp.ml:
```diff
@@ -839,9 +839,9 @@ end = struct
+  let substitute ty =
     match List.assq ty !name_subst with
-    | ty' -> substitute ty'
+    | ty' -> ty'
```
This is a fix, but the commit message seems inexact to me: 
- with separation, the inner substitute always falls because variables on the right side never appear on the left side
- without separation the previous call is mixing the two sides and deciding that  `'a * 'b * int` = `'b * 'c * float`, the left hand side should be displayed as `'c * ' c * int` is not equal to `'c * 'c * float` rather than `'b * 'c * int` is not equal to `'b * 'c * float`.

goldfirere replied:
Without separation, it's even worse: the previous call could loop infinitely. (Indeed I observed this in my branch that does not enforce separation.) Thus my claim that dropping this recursion removes the separation assumption. And, in any case, it seems unnecessary.

Octachron replied:
Yes, I agree that the function call is either unnecessary (with the separation hypothesis) or wrong. I am just not sure I would describe the current call as a "use of the separation assumption".

Octachron reviewed (approved):
The current state looks ready to merge, and I think it is good to simplify preconditions in this part of the typechecker.

gasche commented:
During our discussion today I had the impression that it is not obvious to everyone that `subst` is taken to be *the* relation we are checking on variables, instead of "additional equalities" that are true in addition to reflexivity. It might be useful to have a comment that makes this explicit -- this is implicitly asserted in the comment about `eqtype` below, but in a way that assumes more knowledge of the rest of the file.
